echion/greenlets.h

```diff
@@ -0,0 +1,95 @@
+// This file is part of "echion" which is released under MIT.
+//
+// Copyright (c) 2025 Gabriele N. Tornetta <phoenix1987@gmail.com>.
+
+#pragma once
+
+#include <Python.h>
+#define Py_BUILD_CORE
+
+
+#include <echion/stacks.h>
+#include <echion/strings.h>
+
+
+#define FRAME_NOT_SET Py_False  // Sentinel for frame cell
+
+
+class GreenletInfo
+{
+public:
+    typedef std::unique_ptr<GreenletInfo> Ptr;
+    typedef std::reference_wrapper<GreenletInfo> Ref;
+    typedef uintptr_t ID;
+
+    ID greenlet_id = 0;
+    StringTable::Key name;
+    PyObject* frame_cell = NULL;
+
+    GreenletInfo(ID id, PyObject* frame_cell, StringTable::Key name)
+        : greenlet_id(id), frame_cell(frame_cell), name(name)
+    {
+    }
+
+    int unwind(PyObject*, PyThreadState*, FrameStack&);
+};
+
+// ----------------------------------------------------------------------------
+
+int GreenletInfo::unwind(PyObject* frame, PyThreadState* tstate, FrameStack& stack)
+{
+    PyObject* frame_addr = NULL;
+#if PY_VERSION_HEX >= 0x030d0000
+    frame_addr =
+        frame == Py_None
+            ? (PyObject*)tstate->current_frame
+            : reinterpret_cast<PyObject*>(reinterpret_cast<struct _frame*>(frame)->f_frame);
+#elif PY_VERSION_HEX >= 0x030b0000
+    if (frame == Py_None)
+    {
+        _PyCFrame cframe;
+        _PyCFrame* cframe_addr = tstate->cframe;
+        if (copy_type(cframe_addr, cframe))
+            // TODO: Invalid frame
+            return 0;
+
+        frame_addr = (PyObject*)cframe.current_frame;
+    }
+    else
+    {
+        frame_addr = reinterpret_cast<PyObject*>(reinterpret_cast<struct _frame*>(frame)->f_frame);
+    }
+
+#else  // Python < 3.11
+    frame_addr = frame == Py_None ? (PyObject*)tstate->frame : frame;
+#endif
+    auto count = unwind_frame(frame_addr, stack);
+
+    stack.push_back(Frame::get(name));
+
+    return count + 1;
```

Comment:
Initially wondered about +1, but looked right above and it made sense, as this function is adding a frame for `name`

Reply:
That's right. Maybe best to leave a comment as I'm sure it looks obvious to me now but I'll have no recollection in a month's time 😅